File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/util/game/Artifact.java
package org.firstinspires.ftc.teamcode.util.game;

import androidx.annotation.NonNull;

import java.util.Objects;

public class Artifact {
    // Static instances for clean pattern definition in Motif
    public static final Artifact GREEN = new Artifact(Color.GREEN);
    public static final Artifact PURPLE = new Artifact(Color.PURPLE);
    public static final Artifact NONE = new Artifact(Color.NONE);
    public Color color;

    public Artifact() {
        this.color = Color.NONE;
    }

    public Artifact(Color color) {
        this.color = color;
    }

    @NonNull
    @Override
    public String toString() {
        return "" + color.name().charAt(0);
    }

    /**
     * Determines equality based ONLY on the color, which is essential for Motif
     * pattern comparison.
     */
    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (o == null || getClass() != o.getClass())
            return false;
        Artifact artifact = (Artifact) o;
        return color == artifact.color;
    }

    @Override
    public int hashCode() {
        return Objects.hash(color);
    }

    public enum Color {
        GREEN,
        PURPLE,
        NONE
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/kotlin_mirror/util/game/Artifact.kt
package org.firstinspires.ftc.teamcode.kotlin_mirror.util.game

import java.util.Objects

/**
 * Represents a single game element (Artifact) with a specific color.
 *
 * Used in conjunction with [Motif] and [Classifier] to track game state.
 */
class Artifact(var color: Color = Color.NONE) {

    /**
     * Creates an empty artifact (NONE).
     */
    constructor() : this(Color.NONE)

    override fun toString(): String {
        return "${color.name[0]}"
    }

    /**
     * Checks for equality based solely on the artifact's color.
     *
     * @param other The object to compare.
     * @return true if the colors match, false otherwise.
     */
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val artifact = other as Artifact
        return color == artifact.color
    }

    override fun hashCode(): Int {
        return Objects.hash(color)
    }

    /**
     * Enumeration of possible artifact colors.
     */
    enum class Color {
        GREEN,
        PURPLE,
        NONE
    }

    companion object {
        // Predefined artifact instances for common colors
        @JvmField
        val GREEN = Artifact(Color.GREEN)
        @JvmField
        val PURPLE = Artifact(Color.PURPLE)
        @JvmField
        val NONE = Artifact(Color.NONE)
    }
}
